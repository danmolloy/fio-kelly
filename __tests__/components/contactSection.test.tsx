import "@testing-library/jest-dom"
import { render, screen, fireEvent, act, waitFor } from "@testing-library/react"
import ContactSection from "../../components/contactSection"
import axios from "axios"

const mockPost = jest.spyOn(axios, 'post');


/* jest.mock("axios", () => ({
  axios: {
    post: jest.fn()
  }
})) */

describe("<ContactSection />", () => {
  beforeEach(() => {
    render(<ContactSection />)
  })
  it("contact-form is in the document", () => {
    const contactForm = screen.getByTestId("contact-form")
    expect(contactForm).toBeInTheDocument()
  })
  it("'Contact' title is in the document", () => {
    const contactTitle = screen.getByText("Contact")
    expect(contactTitle).toBeInTheDocument()
  })
  it("MAS link is in the document with help text", () => {
    const contactForm = screen.getByTestId("contact-form")
    expect(contactForm.textContent).toMatch("For concert bookings, please contact The Musicians Answering Service.")
    const masLink = screen.getByText("The Musicians Answering Service")
    expect(masLink).toBeInTheDocument()
    expect(masLink).toHaveAttribute("href", "https://maslink.co.uk/client-directory?client=KELLF1&instrument=FLUTE1")
    expect(masLink).toHaveAttribute("target", "_blank")
  })
  it("general enquiry help text is in the document", () => {
    const contactForm = screen.getByTestId("contact-form")
    expect(contactForm.textContent).toMatch("For all other enquiries, complete the form below.")
  })
  it("name input is in the document with label, type and name attrs", () => {
    const nameInput = screen.getByLabelText("Name")
    expect(nameInput).toBeInTheDocument()
    expect(nameInput).toHaveAttribute("name", "name")
    expect(nameInput).toHaveAttribute("type", "text")
  })
  it("email input is in the document with label, type and name attrs", () => {
    const emailInput = screen.getByLabelText("Email")
    expect(emailInput).toBeInTheDocument()
    expect(emailInput).toHaveAttribute("name", "email")
    expect(emailInput).toHaveAttribute("type", "email")
  })
  it("message input is in the document with label, name, type & maxLength attrs", () => {
    const messageInput = screen.getByLabelText("Message")
    expect(messageInput).toBeInTheDocument()
    expect(messageInput).toHaveAttribute("name", "message")
    expect(messageInput).toHaveAttribute("type", "textarea")
    expect(messageInput).toHaveAttribute("maxLength", "500")
  })
  it("Submit btn is in the document", () => {
    const submitBtn = screen.getByText("Submit")
    expect(submitBtn).toBeInTheDocument()
  })
  it("err messages for name, email and message render on submit btn click", async () => {
    const submitBtn = screen.getByText("Submit")
    await act(async () => 
      await fireEvent.click(submitBtn)
    )
    const nameErr = screen.getByText("name required")
    expect(nameErr).toBeInTheDocument()
    const emailErr = screen.getByText("email required")
    expect(emailErr).toBeInTheDocument()
    const messageErr = screen.getByText("message required")
    expect(messageErr).toBeInTheDocument()
    expect(mockPost).not.toHaveBeenCalled()
  })
  it("calls axios.post(args) on submit with expect 'Sending..' and 'Success' messages", async () => {
    const nameInput = screen.getByLabelText("Name")
    const fakeName = "Joe Bloggs"
    const fakeEmail = "joe@bloggs.com.au"
    const fakeMsg = "Lorem Ipsum"
    act(() => {
      fireEvent.change(nameInput, {target: { value: fakeName}})
    })
    const emailInput = screen.getByLabelText("Email")
    act(() => {
      fireEvent.change(emailInput, {target: { value: fakeEmail}})
    })
    const messageInput = screen.getByLabelText("Message")
    act(() => {
      fireEvent.change(messageInput, {target: { value: fakeMsg}})
    })
    const submitBtn = screen.getByText("Submit")
    await act(async () => {
      await fireEvent.click(submitBtn)
    })
    const contactForm = screen.getByTestId("contact-form")

    expect(contactForm.textContent).toMatch("sending...")
    await act(async () => {
      await new Promise(resolve => setTimeout(resolve, 1000))
    })

    expect(mockPost).toHaveBeenCalledWith("/api/contact", {
      name: fakeName,
      email: fakeEmail,
      message: fakeMsg
    })
    //expect(contactForm.textContent).toMatch("Message recieved!I will get back to you as soon as possible.")

  })
  //it("sendFail message renders if submit throws err", () => {})
  it("matches snapshot", () => {
    const contactForm = screen.getByTestId("contact-form")
    expect(contactForm).toMatchSnapshot()
  })
  
})