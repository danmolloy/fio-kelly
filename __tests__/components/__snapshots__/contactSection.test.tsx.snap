// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ContactSection /> matches snapshot 1`] = `
<div
  class="flex flex-row w-full justify-between  h-[30rem]"
  data-testid="contact-form"
>
  <form
    action="#"
    class="font-sans flex flex-col lg:w-1/2  "
  >
    <div
      class=" mb-2 "
    >
      <p
        class="my-1"
      >
        For concert bookings, please contact
         
        <a
          class="hover:text-blue-500 underline"
          href="https://maslink.co.uk/client-directory?client=KELLF1&instrument=FLUTE1"
          target="_blank"
        >
          The Musicians Answering Service
        </a>
        .
      </p>
      <p
        class="my-1"
      >
        For all other enquiries, complete the form below.
      </p>
    </div>
    <div
      class="flex flex-col my-1"
    >
      <label
        class="form-label"
        for="name-input"
      >
        Name
      </label>
      <input
        class="text-black border border-black rounded-md w-full p-1"
        id="name-input"
        name="name"
        type="text"
        value=""
      />
      <div
        class="h-6"
      />
    </div>
    <div
      class="flex flex-col my-1"
    >
      <label
        class="form-label"
        for="email-input"
      >
        Email
      </label>
      <input
        class="text-black border border-black rounded-md w-full p-1"
        id="email-input"
        name="email"
        type="email"
        value=""
      />
      <div
        class="h-6"
      />
    </div>
    <div
      class="flex flex-col my-1 mb-0"
    >
      <label
        class="form-label"
        for="msg-text"
      >
        Message
      </label>
      <textarea
        class=" text-black border border-black rounded-md w-full p-1"
        id="msg-text"
        maxlength="500"
        multiline="6"
        name="message"
        rows="4"
        type="textarea"
      />
      <div
        class="h-6"
      />
    </div>
    <button
      class="disabled:opacity-40 disabled:hover:cursor-default hover:bg-blue-50 border shadow-sm border-blue-500 text-blue-500 m-2 p-1 w-24 rounded self-end"
      id="submit-button"
      type="submit"
    >
      Submit
    </button>
    <div />
  </form>
  <div
    class="hidden lg:flex mx-12"
  >
    <img
      alt="Fiona Kelly"
      data-nimg="1"
      decoding="async"
      height="425"
      loading="lazy"
      src="/_next/image?url=%2Fimages%2Ffiona-profile.jpg&w=750&q=75"
      srcset="/_next/image?url=%2Fimages%2Ffiona-profile.jpg&w=384&q=75 1x, /_next/image?url=%2Fimages%2Ffiona-profile.jpg&w=750&q=75 2x"
      style="color: transparent;"
      width="325"
    />
  </div>
</div>
`;
